dtm0/service.c

```diff
@@ -453,224 +402,75 @@ M0_INTERNAL bool m0_dtm0_in_ut(void)
+				   " in the conf cache.", FID_P(rem_svc_fid));
+	rem_svc_conf = M0_CONF_CAST(obj, m0_conf_service);
+	obj = m0_conf_obj_grandparent(obj);
+	M0_ASSERT_INFO(obj != NULL, "Service " FID_F " does not belong to "
+		       "any process?", FID_P(rem_svc_fid));
 
-	/** UT workaround */
-	if (!m0_confc_is_inited(confc)) {
-		M0_LOG(M0_WARN, "confc is not initiated!");
-		return M0_RC(0);
-	}
+	rem_proc_conf = M0_CONF_CAST(obj, m0_conf_process);
+	if (rem_svc_conf->cs_type != M0_CST_DTM0)
+		return M0_ERR_INFO(-ENOENT, "Not a DTM0 service.");
 
-	rc = m0_confc_root_open(confc, &root);
-	if (rc != 0)
-		return M0_ERR(rc);
+	dopr_tlink_init(proc);
+	dopr_tlist_add(&dtms->dos_processes, proc);
 
-	rc = m0_conf_diter_init(&it, confc,
-				&root->rt_obj,
-				M0_CONF_ROOT_NODES_FID,
-				M0_CONF_NODE_PROCESSES_FID);
-	if (rc != 0) {
-		m0_confc_close(&root->rt_obj);
-		return M0_ERR(rc);
-	}
+	proc->dop_rproc_fid = rem_proc_conf->pc_obj.co_id;
+	proc->dop_rserv_fid = rem_svc_conf->cs_obj.co_id;
 
-	while ((rc = m0_conf_diter_next_sync(&it, conf_obj_is_process)) > 0) {
-		obj = m0_conf_diter_result(&it);
-		process = M0_CONF_CAST(obj, m0_conf_process);
-		rproc_fid = process->pc_obj.co_id;
-		rc = m0_conf_process2service_get(confc, &rproc_fid,
-						 M0_CST_DTM0, &rserv_fid);
-		if (rc == 0) {
-			if (m0_fid_eq(&rproc_fid, current_proc_fid))
-				/* skip current process */
-				continue;
-			M0_ALLOC_PTR(dtm0_process);
-			M0_ASSERT(dtm0_process != NULL); /* XXX */
-			dtm0_process__ha_state_subscribe(dtm0_process, obj);
-			dopr_tlink_init(dtm0_process);
-			dopr_tlist_add(&s->dos_processes, dtm0_process);
-			dtm0_process->dop_rproc_fid = rproc_fid;
-			dtm0_process->dop_rserv_fid = rserv_fid;
-			dtm0_process->dop_rep =
-				m0_strdup(process->pc_endpoint);
-			dtm0_process->dop_dtm0_service = service;
-		}
-	}
+	proc->dop_rep = m0_strdup(rem_proc_conf->pc_endpoint);
 
-	m0_conf_diter_fini(&it);
-	m0_confc_close(&root->rt_obj);
+	m0_long_lock_init(&proc->dop_llock);
 
 	return M0_RC(0);
 }
 
-static void dtm0_service__ha_unsubscribe(struct m0_reqh_service *reqh_service)
+M0_INTERNAL void dtm0_process_fini(struct dtm0_process *proc)
 {
-	struct dtm0_process    *process;
-	struct m0_dtm0_service *service;
-	int                     rc;
+	dopr_tlink_fini(proc);
+	m0_free(proc->dop_rep);
+	m0_long_lock_fini(&proc->dop_llock);
+}
 
-	M0_PRE(reqh_service != NULL);
-	service = container_of(reqh_service, struct m0_dtm0_service,
-			       dos_generic);
+M0_INTERNAL void dtm0_service_conns_term(struct m0_dtm0_service *service)
+{
+	struct dtm0_process *process;
+	int                  rc;
 
 	M0_ENTRY("dtms=%p", service);
 
 	while ((process = dopr_tlist_pop(&service->dos_processes)) != NULL) {
 		rc = dtm0_process_disconnect(process);
-		M0_ASSERT_INFO(rc == 0, "TODO: Disconnect failures"
-			       " are not handled yet.");
-		dtm0_process__ha_state_unsubscribe(process);
-		dopr_tlink_fini(process);
+		M0_ASSERT_INFO(rc == 0, "Failed to disconnect from %p?",
```

Comment:
Do we really need to assert here?
The remote endpoint may be dead or unreachable.
So, this disconnect() may fail at this moment?

Reply:
We have two expected failures, they are handled separately (timed out and canceled, see `dtm0_process_disconnect`). Any other kind of failure (for example, ENOMEM) is not allowed here because the context (m0_reqh_service_ops::prepare_to_stop) is not able to handle failures. So that, we panicking here instead returning an error code to the caller.

Reply:
OK